specs/networking/libp2p-standardization.md

```diff
@@ -0,0 +1,131 @@
+ETH 2.0 Networking Spec - Libp2p standard protocols
+===
+
+# Abstract
+
+Ethereum 2.0 clients plan to use the libp2p protocol networking stack for
+mainnet release. This document aims to standardize the libp2p client protocols,
+configuration and messaging formats.
+
+# Libp2p Protocols
+
+## Gossipsub
+
+#### Protocol id: `/meshsub/1.0.0`
+
+*Note: Parameters listed here are subject to a large-scale network feasibility
+study*
+
+The [Gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub)
+protocol will be used for block and attestation propagation across the
+network.
+
+### Configuration Parameters
+
+Gossipsub has a number of internal configuration parameters which directly
+effect the network performance.  Clients can implement independently, however
+we aim to standardize these across clients to optimize the gossip network for
+propagation times and message duplication. Current network-related defaults are:
+
+```
+(
+	// The target number of peers in the overlay mesh network (D in the libp2p specs).
+	mesh_size: 6
+	// The minimum number of peers in the mesh network before adding more (D_lo in the libp2p specs).
+	mesh_lo: 4
+	// The maximum number of peers in the mesh network before removing some (D_high in the libp2p sepcs).
+	mesh_high: 12
+	// The number of peers to gossip to during a heartbeat (D_lazy in the libp2p sepcs).
+	gossip_lazy: 6 // defaults to `mesh_size`
+	// Time to live for fanout peers (seconds).
+	fanout_ttl: 60
+	// The number of heartbeats to gossip about.
+	gossip_history: 3
+	// Time between each heartbeat (seconds).
+	heartbeat_interval: 1
+)
+```
+
+### Topics
+
+*The Go and Js implementations use string topics - This is likely to be
+updated to topic hashes in later versions - https://github.com/libp2p/rust-libp2p/issues/473*
+
+For Eth2.0 clients, topics will be sent as `SHA2-256` hashes of the topic string.
+
+There is one dedicated topic for propagating beacon blocks and aggregated
+attestations across the network. This topic will have the string
+`beacon_chain`. Each shard will have it's own topic allowing relevant parties
+to subscribe to in order to receive local shard attestations. The shard topics are
+prefixed with `shard` followed by the number of the shard. For example,
+messages relating to shard 10, will have the topic string `shard10`.
```

Comment:
We might want to launch with fewer shard subnets than shards to promote stability in networks and reduce complexity. To do so, we would introduce a constant `SHARD_SUBNET_COUNT` and shards would be assigned to `shard{shard_number % SHARD_SUBNET_COUNT)`.

If we launch the beacon chain with no notion of "persistent committees", then this will very likely need to be a requirement, otherwise each shard subnet would get almost entirely dropped each epoch.

Reply:
Yep sure. I'll add this in.